## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 513995`
Commit: [`ab3c38bdf5f2a3092e43bc1badb523f29a83c2f2`](https://github.com/NixOS/nixpkgs/commit/ab3c38bdf5f2a3092e43bc1badb523f29a83c2f2) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/ab3c38bdf5f2a3092e43bc1badb523f29a83c2f2..pull/513995/head))
Merge: [`90dff507f6fb9f708b8d50c4eebd4fe75d9091a7`](https://github.com/NixOS/nixpkgs/commit/90dff507f6fb9f708b8d50c4eebd4fe75d9091a7)

Logs: https://github.com/06kellyjac/nixpkgs-review-gha/actions/runs/24996850581


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>grummage</li>
    <li>grummage.dist</li>
    <li>grype</li>
    <li>sbomnix</li>
    <li>sbomnix.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>grummage</li>
    <li>grummage.dist</li>
    <li>grype</li>
    <li>sbomnix</li>
    <li>sbomnix.dist</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 5 packages failed to build:</summary>
  <ul>
    <li>grummage</li>
    <li>grummage.dist</li>
    <li>grype</li>
    <li>sbomnix</li>
    <li>sbomnix.dist</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>grype</summary>
<pre>--- FAIL: Test_defaultHTTPClientHasCert (0.81s)
    --- FAIL: Test_defaultHTTPClientHasCert/should_use_single_custom_cert (0.81s)
        curator_test.go:137: Generating Key/Cert Fixture
        curator_test.go:165: out: ./generate-x509-cert-pair.sh
        curator_test.go:165: err: ++ hostname
        curator_test.go:165: err: ++ sed &#x27;s/[&#x27;\&#x27;&#x27;&#x27;\&#x27;&#x27;]/&#x27;\&#x27;&#x27;/g&#x27;
        curator_test.go:165: err: ++ sed &#x27;s/[^a-zA-Z0-9.-]/-/g&#x27;
        curator_test.go:165: err: + HOSTNAME=sat12-dp146-42fc911f-be07-4fc5-8bf5-5ed16efd1274-1A0D5F27603C.local
        curator_test.go:165: err: + openssl genrsa -out server.key 2048
        curator_test.go:165: err: + openssl req -new -x509 -sha256 -key server.key -out server.crt -days 3650 -reqexts SAN -extensions SAN -config /dev/fd/63 -subj /C=US/ST=Test/L=Test/O=Test/CN=sat12-dp146-42fc911f-be07-4fc5-8bf5-5ed16efd1274-1A0D5F27603C.local.local
        curator_test.go:165: err: ++ cat /etc/ssl/openssl.cnf /dev/fd/63
        curator_test.go:165: err: +++ printf &#x27;[SAN]\nsubjectAltName=DNS:sat12-dp146-42fc911f-be07-4fc5-8bf5-5ed16efd1274-1A0D5F27603C.local.local&#x27;
        curator_test.go:165: err: cat: /etc/ssl/openssl.cnf: Operation not permitted
        curator_test.go:165: err: 0092DD0202000000:error:06800097:asn1 encoding routines:ASN1_mbstring_ncopy:string too long:crypto/asn1/a_mbstr.c:115:maxsize=64
        curator_test.go:165: err: req: Error adding subject name attribute &quot;/CN=sat12-dp146-42fc911f-be07-4fc5-8bf5-5ed16efd1274-1A0D5F27603C.local.local&quot;
        curator_test.go:165: err: make: *** [Makefile:25: server.crt] Error 1
        curator_test.go:181: failed to generate fixture: rc=2
FAIL
FAIL	github.com/anchore/grype/grype/db/v5/distribution	5.661s
FAIL</pre>
</details>
</details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 5 packages failed to build:</summary>
  <ul>
    <li>grummage</li>
    <li>grummage.dist</li>
    <li>grype</li>
    <li>sbomnix</li>
    <li>sbomnix.dist</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>grype</summary>
<pre>--- FAIL: Test_defaultHTTPClientHasCert (0.09s)
    --- FAIL: Test_defaultHTTPClientHasCert/should_use_single_custom_cert (0.09s)
        curator_test.go:137: Generating Key/Cert Fixture
        curator_test.go:165: out: ./generate-x509-cert-pair.sh
        curator_test.go:165: err: ++ hostname
        curator_test.go:165: err: ++ sed &#x27;s/[&#x27;\&#x27;&#x27;&#x27;\&#x27;&#x27;]/&#x27;\&#x27;&#x27;/g&#x27;
        curator_test.go:165: err: ++ sed &#x27;s/[^a-zA-Z0-9.-]/-/g&#x27;
        curator_test.go:165: err: + HOSTNAME=sjc20-bb660-ae510f5e-0a65-4d0f-9aa8-ae13c9798a8f-C24B20E0A703.local
        curator_test.go:165: err: + openssl genrsa -out server.key 2048
        curator_test.go:165: err: + openssl req -new -x509 -sha256 -key server.key -out server.crt -days 3650 -reqexts SAN -extensions SAN -config /dev/fd/63 -subj /C=US/ST=Test/L=Test/O=Test/CN=sjc20-bb660-ae510f5e-0a65-4d0f-9aa8-ae13c9798a8f-C24B20E0A703.local.local
        curator_test.go:165: err: ++ cat /etc/ssl/openssl.cnf /dev/fd/63
        curator_test.go:165: err: +++ printf &#x27;[SAN]\nsubjectAltName=DNS:sjc20-bb660-ae510f5e-0a65-4d0f-9aa8-ae13c9798a8f-C24B20E0A703.local.local&#x27;
        curator_test.go:165: err: cat: /etc/ssl/openssl.cnf: Operation not permitted
        curator_test.go:165: err: 40A29BFF01000000:error:06800097:asn1 encoding routines:ASN1_mbstring_ncopy:string too long:crypto/asn1/a_mbstr.c:115:maxsize=64
        curator_test.go:165: err: req: Error adding subject name attribute &quot;/CN=sjc20-bb660-ae510f5e-0a65-4d0f-9aa8-ae13c9798a8f-C24B20E0A703.local.local&quot;
        curator_test.go:165: err: make: *** [Makefile:25: server.crt] Error 1
        curator_test.go:181: failed to generate fixture: rc=2
FAIL
FAIL	github.com/anchore/grype/grype/db/v5/distribution	1.796s
FAIL</pre>
</details>
</details>

--- a/pkgs/by-name/gr/grype/0001-test_helpers-redact-support-nix.patch
+++ b/pkgs/by-name/gr/grype/0001-test_helpers-redact-support-nix.patch
@@ -1,0 +1,35 @@
+diff --git a/grype/presenter/internal/test_helpers.go b/grype/presenter/internal/test_helpers.go
+index 2bfafbc2..6e70ed87 100644
+--- a/grype/presenter/internal/test_helpers.go
++++ b/grype/presenter/internal/test_helpers.go
+@@ -1,6 +1,7 @@
+ package internal
+ 
+ import (
++	"os"
+ 	"regexp"
+ 	"testing"
+ 
+@@ -92,8 +93,21 @@ func Redact(s []byte) []byte {
+ 	cycloneDxBomRefPattern := regexp.MustCompile(`[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`)
+ 	tempDirPattern := regexp.MustCompile(`/tmp/[^"]+`)
+ 	macTempDirPattern := regexp.MustCompile(`/var/folders/[^"]+`)
++	// nix build dir isn't always /build so we'll identify it by $TMP
++	nixDir := os.Getenv("TMP")
++	if nixDir == "" {
++		panic("missing $TMP for identifying nix build dir")
++	}
++	end := regexp.QuoteMeta(nixDir)
++	pattern := `[^"]+`
++	// add slash prefix to pattern only if required
++	// don't use strings lib to reduce chance of merge conflicts above
++	if end[len(end)-1] != '/' {
++		pattern = "/" + pattern
++	}
++	nixBuildDirPattern := regexp.MustCompile(end + pattern)
+ 
+-	for _, pattern := range []*regexp.Regexp{serialPattern, rfc3339Pattern, refPattern, uuidPattern, cycloneDxBomRefPattern, tempDirPattern, macTempDirPattern} {
++	for _, pattern := range []*regexp.Regexp{serialPattern, rfc3339Pattern, refPattern, uuidPattern, cycloneDxBomRefPattern, tempDirPattern, macTempDirPattern, nixBuildDirPattern} {
+ 		s = pattern.ReplaceAll(s, []byte(""))
+ 	}
+ 	return s

--- a/pkgs/by-name/gr/grype/package.nix
+++ b/pkgs/by-name/gr/grype/package.nix
@@ -14,6 +14,9 @@ buildGoModule (finalAttrs: {
   pname = "grype";
   version = "0.111.1";
 
+  # required for tests
+  __darwinAllowLocalNetworking = true;
+
   src = fetchFromGitHub {
     owner = "anchore";
     repo = "grype";
@@ -34,6 +37,12 @@ buildGoModule (finalAttrs: {
   proxyVendor = true;
 
   vendorHash = "sha256-rsdZt+xKjIJpWS5pYx8A+ryY1D2WIKquKjsQBkxToUQ=";
+
+  patches = [
+    # several test golden files have unstable paths based on the platform
+    # this patch adjusts the `Redact` helper to also work for builds by nix.
+    ./0001-test_helpers-redact-support-nix.patch
+  ];
 
   nativeBuildInputs = [ installShellFiles ];
 
@@ -73,52 +82,37 @@ buildGoModule (finalAttrs: {
 
     # patch utility script
     patchShebangs grype/db/v5/distribution/testdata/tls/generate-x509-cert-pair.sh
-
-    # FIXME: these tests fail when building with Nix
-    substituteInPlace test/cli/config_test.go \
-      --replace-fail "Test_configLoading" "Skip_configLoading"
-    substituteInPlace test/cli/db_providers_test.go \
-      --replace-fail "TestDBProviders" "SkipDBProviders"
-    substituteInPlace grype/presenter/cyclonedx/presenter_test.go \
-      --replace-fail "TestCycloneDxPresenterDir" "SkipCycloneDxPresenterDir" \
-      --replace-fail "Test_CycloneDX_Valid" "Skip_CycloneDX_Valid"
-
-    # remove tests that depend on docker
-    substituteInPlace test/cli/cmd_test.go \
-      --replace-fail "TestCmd" "SkipCmd"
-    substituteInPlace grype/pkg/provider_test.go \
-      --replace-fail "TestSyftLocationExcludes" "SkipSyftLocationExcludes"
-    substituteInPlace test/cli/cmd_test.go \
-      --replace-fail "Test_descriptorNameAndVersionSet" "Skip_descriptorNameAndVersionSet"
-
-    # remove tests that depend on git
-    substituteInPlace test/cli/db_validations_test.go \
-      --replace-fail "TestDBValidations" "SkipDBValidations"
-    substituteInPlace test/cli/registry_auth_test.go \
-      --replace-fail "TestRegistryAuth" "SkipRegistryAuth"
-    substituteInPlace test/cli/sbom_input_test.go \
-      --replace-fail "TestSBOMInput_FromStdin" "SkipSBOMInput_FromStdin" \
-      --replace-fail "TestSBOMInput_AsArgument" "SkipSBOMInput_AsArgument"
-    substituteInPlace test/cli/subprocess_test.go \
-      --replace-fail "TestSubprocessStdin" "SkipSubprocessStdin"
-    substituteInPlace grype/internal/packagemetadata/names_test.go \
-      --replace-fail "TestAllNames" "SkipAllNames"
-    substituteInPlace test/cli/version_cmd_test.go \
-      --replace-fail "TestVersionCmdPrintsToStdout" "SkipVersionCmdPrintsToStdout"
-    substituteInPlace grype/presenter/sarif/presenter_test.go \
-      --replace-fail "Test_SarifIsValid" "SkipTest_SarifIsValid"
-    substituteInPlace test/cli/config_test.go \
-      --replace-fail "Test_dpkgUseCPEsForEOLEnvVar" "SkipTest_dpkgUseCPEsForEOLEnvVar" \
-      --replace-fail "Test_rpmUseCPEsForEOLEnvVar" "SkipTest_rpmUseCPEsForEOLEnvVar"
-
-    # May fail on NixOS, probably due bug in how syft handles tmpfs.
-    # See https://github.com/anchore/grype/issues/1822
-    substituteInPlace grype/distro/distro_test.go \
-      --replace-fail "Test_NewDistroFromRelease_Coverage" "SkipTest_NewDistroFromRelease_Coverage"
-
-    # segfault
-    rm grype/db/v5/namespace/cpe/namespace_test.go
   '';
+
+  checkFlags =
+    let
+      skippedTests = [
+        # depend on docker
+        "TestCmd"
+        "TestSyftLocationExcludes"
+        "Test_descriptorNameAndVersionSet"
+        # depend on .git
+        "Test_configLoading"
+        "TestDBProviders"
+        "TestDBValidations"
+        "TestRegistryAuth"
+        "TestRegistryAuthRedactions"
+        "TestSBOMInput_FromStdin"
+        "TestSBOMInput_AsArgument"
+        "TestSubprocessStdin"
+        "TestAllNames"
+        "TestVersionCmdPrintsToStdout"
+        "Test_SarifIsValid"
+        "Test_dpkgUseCPEsForEOLEnvVar"
+        "Test_rpmUseCPEsForEOLEnvVar"
+      ]
+      ++ lib.optionals stdenv.isDarwin [
+        # fails to generate x509 certificate
+        # cat: /etc/ssl/openssl.cnf: Operation not permitted
+        "Test_defaultHTTPClientHasCert"
+      ];
+    in
+    [ "-skip=^${builtins.concatStringsSep "$|^" skippedTests}$" ];
 
   postInstall = lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
     installShellCompletion --cmd grype \

